cudamapper/src/index_gpu_two_indices.cuh

```diff
@@ -0,0 +1,439 @@
+    CGA_CU_CHECK_ERR(cudaMemcpy(read_id_to_basepairs_section_d.data(),
+                                read_id_to_basepairs_section_h.data(),
+                                read_id_to_basepairs_section_h.size() * sizeof(decltype(read_id_to_basepairs_section_h)::value_type),
+                                cudaMemcpyHostToDevice));
+
+    CGA_LOG_INFO("Allocating {} bytes for merged_basepairs_d", merged_basepairs_h.size() * sizeof(decltype(merged_basepairs_h)::value_type));
+    device_buffer<decltype(merged_basepairs_h)::value_type> merged_basepairs_d(merged_basepairs_h.size());
+    CGA_CU_CHECK_ERR(cudaMemcpy(merged_basepairs_d.data(),
+                                merged_basepairs_h.data(),
+                                merged_basepairs_h.size() * sizeof(decltype(merged_basepairs_h)::value_type),
+                                cudaMemcpyHostToDevice));
+    merged_basepairs_h.clear();
+    merged_basepairs_h.shrink_to_fit();
+
+    // sketch elements get generated here
+    auto sketch_elements                                                      = SketchElementImpl::generate_sketch_elements(number_of_reads_,
+                                                                       kmer_size_,
+                                                                       window_size_,
+                                                                       first_read_id,
+                                                                       merged_basepairs_d,
+                                                                       read_id_to_basepairs_section_h,
+                                                                       read_id_to_basepairs_section_d);
+    device_buffer<representation_t> representations_d                         = std::move(sketch_elements.representations_d);
+    device_buffer<typename SketchElementImpl::ReadidPositionDirection> rest_d = std::move(sketch_elements.rest_d);
+
+    CGA_LOG_INFO("Deallocating {} bytes from read_id_to_basepairs_section_d", read_id_to_basepairs_section_d.size() * sizeof(decltype(read_id_to_basepairs_section_d)::value_type));
+    read_id_to_basepairs_section_d.free();
+    CGA_LOG_INFO("Deallocating {} bytes from merged_basepairs_d", merged_basepairs_d.size() * sizeof(decltype(merged_basepairs_d)::value_type));
+    merged_basepairs_d.free();
+
+    // *** sort sketch elements by representation ***
+    // As this is a stable sort and the data was initailly grouper by read_id this means that the sketch elements within each representations are sorted by read_id
+    thrust::stable_sort_by_key(thrust::device,
+                               representations_d.data(),
+                               representations_d.data() + representations_d.size(),
+                               rest_d.data());
+
+    // copy the data to member functions (depending on the interface desing these copies might not be needed)
+    representations_d_.resize(representations_d.size());
```

Comment:
It's a bad/outdated interface design. `SketchElementImpl::generate_sketch_elements` returns a `device_buffer` while `IndexGPUTwoIndices` uses a `thrust::device_vector`, so it's not possible to simply move the data.
It might to be possible to do some trickery with raw pointers, but the interface of `SketchElementImpl::generate_sketch_elements` will anyway have to improved so leaving it as it is for now